stan-dev/stan
# Feature/issue 1635 array expressions

#### Submisison Checklist


#### Summary

Adds expressions which denote the contents of an array, per https://github.com/stan-dev/stan/issues/1635

#### Intended Effect

Programs can specify array contents explicitly.

#### How to Verify

Directory stan/src/test-models  contains models plus data files:

*  stan/src/test/test-models/validate_array_expr_primitives.stan
*  stan/src/test/test-models/validate_array_expr_containers.stan
*  stan/src/test/test-models/validate_array_expr_primitives.data.R
*  stan/src/test/test-models/validate_array_expr_containers.data.R

Models should compile and run - example invocation via cmdstan:

```
>  make stan/src/test/test-models/good/validate_array_expr_containers

--- Translating Stan model to C++ code ---
bin/stanc stan/src/test/test-models/good/validate_array_expr_containers.stan --o=stan/src/test/test-models/good/validate_array_expr_containers.hpp
Model name=validate_array_expr_containers_model
Input file=stan/src/test/test-models/good/validate_array_expr_containers.stan
Output file=stan/src/test/test-models/good/validate_array_expr_containers.hpp

--- Linking C++ model ---
clang++ -I src -I stan/src -isystem stan/lib/stan_math/ -isystem stan/lib/stan_math/lib/eigen_3.2.9 -isystem stan/lib/stan_math/lib/boost_1.62.0 -isystem stan/lib/stan_math/lib/cvodes_2.9.0/include -Wall -DEIGEN_NO_DEBUG  -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DFUSION_MAX_VECTOR_SIZE=12 -DNO_FPRINTF_OUTPUT -pipe  -Wno-unused-function -ftemplate-depth-256    -O0 -o stan/src/test/test-models/good/validate_array_expr_containers src/cmdstan/main.cpp -include stan/src/test/test-models/good/validate_array_expr_containers.hpp stan/lib/stan_math/lib/cvodes_2.9.0/lib/libsundials_nvecserial.a stan/lib/stan_math/lib/cvodes_2.9.0/lib/libsundials_cvodes.a 

> stan/src/test/test-models/good/validate_array_expr_containers sample num_samples=1 num_warmup=0 data file=stan/src/test/test-models/good/validate_array_expr_containers.data.R 

....

```


#### Side Effects

#### Documentation

to be written (Bob Carpenter)

#### Reviewer Suggestions

Bob Carpenter, Daniel Lee

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Mitzi Morris

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)
